src/transformation/transformation.ts

```diff
@@ -3,9 +3,305 @@ import {
+
+        const prefix = isClassType(type) ? `${getClassTypeName(type)}#` : "";
+        this.children.set(prefix + name, new ChildNodeFactory(prefix + name, get, set));
+        return this;
+    }
+
+    withFinalizer = function (finalizer: (Output) => void) {
+        this.finalizer = finalizer;
+    }
+
+    getter : (Source) => any = function(path: string) {
+        return function(src: Source) {
+            let sub = src;
+
+            for (const elem in path.split(".")) {
+                if (sub == null)
+                    break;
+                sub = this.getSubExpression(sub, elem);
+            }
+
+            return sub;
+        }
+    }
+
+    private getSubExpression : any | undefined = function (src: any, elem: string) {
+        if (Array.isArray(src)) {
+            return src.map(it => this.getSubExpression(it!, elem));
+        } else {
+            const sourcePropName : string | undefined = Object.keys(src).find(e => e == elem);
+
+            if (!sourcePropName)
+                throw new Error(`${elem} not found in ${src} (class: ${Object.getPrototypeOf(src).constructor.name})`)
+
+            const sourceProp = src[sourcePropName];
+            return sourceProp instanceof Function ? sourceProp() : sourceProp;
+        }
+    }
+}
+
+export class ChildNodeFactory<Source, Target, Child> {
+    constructor(
+        public name: string,
+        public get: (Source) => any | undefined,
+        public setter: (Target, Child?) => void
+    ) {}
+
+    set = function(node: Target, child?: Child) {
+        try {
+            this.setter(node, child);
+        } catch (e) {
+            // TODO: pass e as the cause of this error
+            throw Error(`${this.name} could not set child ${child} of ${node} using ${this.setter}`);
+        }
+    }
+}
+
+/**
+ * Sentinel value used to represent the information that a given property is not a child node.
+ */
+const NO_CHILD_NODE = new ChildNodeFactory<any, any, any>(
+    "",
+    (node) => node,
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    (target, child) => {}
+);
+
+/**
+ * Implementation of a tree-to-tree transformation. For each source node type, we can register a factory that knows how
+ * to create a transformed node. Then, this transformer can read metadata in the transformed node to recursively
+ * transform and assign children.
+ * If no factory is provided for a source node type, a GenericNode is created, and the processing of the subtree stops
+ * there.
+ */
+export class ASTTransformer {
+    private readonly _issues: Issue[];
+    private readonly allowGenericNode: boolean;
+
+    get issues() : Issue[] {
+        return this._issues;
+    }
+
+    /**
+     * Factories that map from source tree node to target tree node.
+     */
+    private factories = new Map<any, NodeFactory<any, any>>();
+    private knownClasses = new Map<string, any>(); //<-- TODO: not sure about this piece of the port from Kolasu
```

Comment:
This is only used to save some typing/boilerplate when building an Ecore metamodel. It's not a core piece of functionality and if it causes any issues I'd say we can safely omit it.